execution/nodeInterface/NodeInterface.go

```diff
@@ -53,90 +51,126 @@ var merkleTopic common.Hash
 	block := n.backend.ChainConfig().ArbitrumChainParams.GenesisBlockNum
 	return arbmath.UintToBig(block), nil
 }
 
+// bool will be false but no error if behind genesis
+func (n NodeInterface) blockNumToMessageIndex(blockNum uint64) (arbutil.MessageIndex, bool, error) {
+	node, err := gethExecFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	blockchain, err := blockchainFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	if blockNum < blockchain.Config().ArbitrumChainParams.GenesisBlockNum {
+		return 0, true, nil
+	}
+	msgIndex, err := node.ExecEngine.BlockNumberToMessageIndex(blockNum)
+	if err != nil {
+		return 0, false, err
+	}
+	return msgIndex, true, nil
+}
+
+func (n NodeInterface) msgNumToInboxBatch(msgIndex arbutil.MessageIndex) (uint64, bool, error) {
+	node, err := gethExecFromNodeInterfaceBackend(n.backend)
+	if err != nil {
+		return 0, false, err
+	}
+	fetcher := node.ExecEngine.GetBatchFetcher()
+	if fetcher == nil {
+		return 0, false, errors.New("batch fetcher not set")
+	}
+	return fetcher.FindInboxBatchContainingMessage(msgIndex)
+}
+
 func (n NodeInterface) FindBatchContainingBlock(c ctx, evm mech, blockNum uint64) (uint64, error) {
-	node, err := arbNodeFromNodeInterfaceBackend(n.backend)
+	msgIndex, found, err := n.blockNumToMessageIndex(blockNum)
 	if err != nil {
 		return 0, err
 	}
-	return findBatchContainingBlock(node, node.TxStreamer.GenesisBlockNumber(), blockNum)
+	if !found {
+		return 0, fmt.Errorf("block %v is part of genesis", blockNum)
+	}
+	res, found, err := n.msgNumToInboxBatch(msgIndex)
+	if err == nil && !found {
+		return 0, errors.New("block not yet found on any block")
+	}
+	return res, err
 }
 
 func (n NodeInterface) GetL1Confirmations(c ctx, evm mech, blockHash bytes32) (uint64, error) {
-	node, err := arbNodeFromNodeInterfaceBackend(n.backend)
+	node, err := gethExecFromNodeInterfaceBackend(n.backend)
 	if err != nil {
 		return 0, err
 	}
-	if node.InboxReader == nil {
-		return 0, nil
-	}
-	bc, err := blockchainFromNodeInterfaceBackend(n.backend)
+	blockchain, err := blockchainFromNodeInterfaceBackend(n.backend)
 	if err != nil {
 		return 0, err
 	}
-	header := bc.GetHeaderByHash(blockHash)
+	header := blockchain.GetHeaderByHash(blockHash)
 	if header == nil {
 		return 0, errors.New("unknown block hash")
 	}
 	blockNum := header.Number.Uint64()
-	genesis := node.TxStreamer.GenesisBlockNumber()
-	batch, err := findBatchContainingBlock(node, genesis, blockNum)
+
+	// blocks behind genesis are treated as belonging to batch 0
+	msgNum, _, err := n.blockNumToMessageIndex(blockNum)
 	if err != nil {
-		if errors.Is(err, blockInGenesis) {
-			batch = 0
-		} else if errors.Is(err, blockAfterLatestBatch) {
-			return 0, nil
-		} else {
-			return 0, err
-		}
+		return 0, err
+	}
+	// batches not yet posted have 0 confirmations but no error
+	batchNum, found, err := n.msgNumToInboxBatch(msgNum)
+	if err != nil || !found {
+		return 0, err
```

Comment:
I'd prefer if we split out this into two if statements. I know it's not as nice but there's been quite a few times where I've changed the return to something like `return 0, fmt.Errorf("msgNumToInboxBatch failed: %w", err)` and now it'll return a non-nil error if !found. errors.Wrap was nice for this in the past when we were using that library.